pkg/jsonnet/importer.go

```diff
@@ -1,67 +1,122 @@
 package jsonnet
 
 import (
-	"bytes"
 	"encoding/json"
 	"io"
 	"path/filepath"
+	"strings"
 
 	jsonnet "github.com/google/go-jsonnet"
 	"github.com/pkg/errors"
 	"gopkg.in/yaml.v3"
 )
 
+const locationInternal = "<internal>"
+
 // ExtendedImporter wraps jsonnet.FileImporter to add additional functionality:
 // - `import "file.yaml"`
+// - `import "tk"`
 type ExtendedImporter struct {
-	fi *jsonnet.FileImporter
+	fi           *jsonnet.FileImporter
+	interceptors []ImportInterceptor
+	processors   []ImportProcessor
 }
 
+// ImportInterceptor are executed before the actual importing. If they return
+// something, this value is used.
+type ImportInterceptor func(importedFrom, importedPath string) (c *string, foundAt string, err error)
+
+// ImportProcessor are executed after the file import and may modify the result
+// further
+type ImportProcessor func(contents, foundAt string) (c *string, err error)
+
 // NewExtendedImporter returns a new instance of ExtendedImporter with the
 // correct jpaths set up
 func NewExtendedImporter(jpath []string) *ExtendedImporter {
 	return &ExtendedImporter{
 		fi: &jsonnet.FileImporter{
 			JPaths: jpath,
 		},
+		processors:   []ImportProcessor{yamlProcessor},
+		interceptors: []ImportInterceptor{tkInterceptor},
 	}
 }
 
 // Import implements the functionality offered by the ExtendedImporter
 func (i *ExtendedImporter) Import(importedFrom, importedPath string) (contents jsonnet.Contents, foundAt string, err error) {
+	// check if an interceptor handles this
+	for _, interceptor := range i.interceptors {
+		c, foundAt, err := interceptor(importedFrom, importedPath)
+		switch {
+		case err != nil:
+			return jsonnet.Contents{}, "", err
+		case c == nil:
+			continue
+		default:
+			return jsonnet.MakeContents(*c), foundAt, nil
+		}
+	}
+
 	// regularly import
 	contents, foundAt, err = i.fi.Import(importedFrom, importedPath)
 	if err != nil {
```

Comment:
Why wouldn't importing "tk" return an error here (i.e., before it every reaches the processor)?

Reply:
"tk" is an interceptor

Reply:
"yaml" is a processor

Reply:
Gotcha, my bad.